beacon/light/committee_chain.go

```diff
@@ -0,0 +1,587 @@
+			if len(enc) != common.HashLength {
+				return common.Hash{}, errors.New("incorrect length for committee root entry in the database")
+			}
+			return common.BytesToHash(enc), nil
+		}),
+		committees: newCanonicalStore[*types.SerializedSyncCommittee](db, rawdb.SyncCommitteeKey, func(committee *types.SerializedSyncCommittee) ([]byte, error) {
+			return committee[:], nil
+		}, func(enc []byte) (*types.SerializedSyncCommittee, error) {
+			if len(enc) == types.SerializedSyncCommitteeSize {
+				committee := new(types.SerializedSyncCommittee)
+				copy(committee[:], enc)
+				return committee, nil
+			}
+			return nil, errors.New("incorrect length for serialized committee entry in the database")
+		}),
+		updates: newCanonicalStore[*types.LightClientUpdate](db, rawdb.BestUpdateKey, func(update *types.LightClientUpdate) ([]byte, error) {
+			return rlp.EncodeToBytes(update)
+		}, func(enc []byte) (*types.LightClientUpdate, error) {
+			update := new(types.LightClientUpdate)
+			if err := rlp.DecodeBytes(enc, update); err != nil {
+				return nil, err
+			}
+			return update, nil
+		}),
+		syncCommitteeCache: lru.NewCache[uint64, syncCommittee](10),
+		db:                 db,
+		sigVerifier:        sigVerifier,
+		clock:              clock,
+		unixNano:           unixNano,
+		config:             config,
+		signerThreshold:    signerThreshold,
+		enforceTime:        enforceTime,
+		minimumUpdateScore: types.UpdateScore{
+			SignerCount:    uint32(signerThreshold),
+			SubPeriodIndex: params.SyncPeriodLength / 16,
+		},
+	}
+
+	// check validity constraints
+	if !s.updates.periods.IsEmpty() {
+		if s.fixedRoots.periods.IsEmpty() || s.updates.periods.First < s.fixedRoots.periods.First ||
+			s.updates.periods.First >= s.fixedRoots.periods.AfterLast {
+			log.Error("Inconsistent database error: first update is not in the fixed roots range")
+		}
+		if s.committees.periods.First > s.updates.periods.First || s.committees.periods.AfterLast <= s.updates.periods.AfterLast {
+			log.Error("Inconsistent database error: missing committees in update range")
+		}
+	}
+	if !s.committees.periods.IsEmpty() {
+		if s.fixedRoots.periods.IsEmpty() || s.committees.periods.First < s.fixedRoots.periods.First ||
+			s.committees.periods.First >= s.fixedRoots.periods.AfterLast {
+			log.Error("Inconsistent database error: first committee is not in the fixed roots range")
+		}
+		if s.committees.periods.AfterLast > s.fixedRoots.periods.AfterLast && s.committees.periods.AfterLast > s.updates.periods.AfterLast+1 {
+			log.Error("Inconsistent database error: last committee is neither in the fixed roots range nor proven by updates")
+		}
+		log.Trace("Sync committee chain loaded", "first period", s.committees.periods.First, "last period", s.committees.periods.AfterLast-1)
+	}
+	// roll back invalid updates (might be necessary if forks have been changed since last time)
+	var batch ethdb.Batch
+	for !s.updates.periods.IsEmpty() {
```

Comment:
I know next to nothing -- is the `for` looping here bounded reasonably? Or can it happen that we shove too much stuff into the batch and it goes OOM?

Reply:
The same applies to any rollback scenario as this just rolls the chain back period by period. I'm not sure how much is too much (this is 3 db items per period or about 1000 per year) but actually rollbacks do not need to be single batched so I changed rollback to only batch deletions of data belonging to a single period (max 3 items).